rest_server/IOMgrConnection.go

```diff
@@ -0,0 +1,462 @@
+// Copyright (C) 2019-2020 Siodb GmbH. All rights reserved.
+// Use of this source code is governed by a license that can be found
+// in the LICENSE file.
+
+package main
+
+import (
+	"SiodbIomgrProtocol"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/golang/protobuf/proto"
+)
+
+var (
+	IOMgrChunkMaxSize    uint32 = (2 * 1024 * 1024 * 1024) - 1
+	MessageLengthMaxSize uint32 = 1 * 1024 * 1024
+)
+
+var (
+	DATABASEENGINERESPONSE    uint32 = 4
+	DATABASEENGINERESTREQUEST uint32 = 13
+)
+
+type IOMgrConnection struct {
+	*TrackedNetConn
+	pool *IOMgrConnPool
+}
+
+func (IOMgrConn *IOMgrConnection) writeJSONPayload(requestID uint64, c *gin.Context) (err error) {
+
+	// Read Payload up to max json payload size
+	var bytesRead int = 0
+	var bytesReadTotal uint64 = 0
+	var bytesWritten uint64 = 0
+	var bytesWrittenTotal uint64 = 0
+	buffer := make([]byte, restServerConfig.RequestPayloadBufferSize)
+
+	siodbLoggerPool.Debug("writeJSONPayload | IOMgrConn.pool.maxJsonPayloadSize: %v, restServerConfig.RequestPayloadBufferSize: %v",
+		IOMgrConn.pool.maxJsonPayloadSize, restServerConfig.RequestPayloadBufferSize)
+
+	for {
+		if bytesReadTotal > uint64(IOMgrConn.pool.maxJsonPayloadSize) {
+			IOMgrConn.Close()
+			IOMgrConn.TrackedNetConn.Conn = nil
+			return fmt.Errorf("JSON payload is too large:  received %v bytes, but expecting at most %v bytes",
+				bytesReadTotal, IOMgrConn.pool.maxJsonPayloadSize)
+		}
+		bytesRead, err = io.ReadFull(c.Request.Body, buffer)
+		if err == io.ErrUnexpectedEOF {
+			bytesWritten, err = IOMgrConn.writeJSONPayloadChunk(buffer, bytesRead)
+			if err != nil {
+				return err
+			}
+			bytesReadTotal += uint64(bytesRead)
+			bytesWrittenTotal += bytesWritten
+			siodbLoggerPool.Debug("writeJSONPayload | bytes Read: %v (total: %v), bytes Written: %v (total: %v)",
+				bytesRead, bytesReadTotal, bytesWritten, bytesWrittenTotal)
+			break
+		} else if err != nil {
+			return err
+		}
+		bytesWritten, err = IOMgrConn.writeJSONPayloadChunk(buffer, bytesRead)
+		if err != nil {
+			return err
+		}
+		bytesReadTotal += uint64(bytesRead)
+		bytesWrittenTotal += bytesWritten
+		siodbLoggerPool.Debug("writeJSONPayload | bytes Read: %v (total: %v), bytes Written: %v (total: %v)",
+			bytesRead, bytesReadTotal, bytesWritten, bytesWrittenTotal)
+	}
+	_, err = IOMgrConn.writeJSONPayloadChunk(buffer, 0)
+	if err != nil {
+		return err
+	}
+
+	// Get Returned message
+	if err = IOMgrConn.readIOMgrResponse(requestID); err != nil {
```

Comment:
Why reading message comes here? This function is called "writeJsonPayload", so the only thing it should do is to write JSON payload. It should not read response. If you want it this way, please rename it to something like "writeJsonPayloadAndReadIOMgrResponse"